src/proof/retainer.rs

```diff
@@ -1,8 +1,51 @@
-use crate::{Nibbles, proof::ProofNodes};
+use crate::{Nibbles, TrieMask, proof::ProofNodes};
 use alloy_primitives::Bytes;
+use alloy_rlp::EMPTY_STRING_CODE;
 
 use alloc::vec::Vec;
 
+/// Tracks various datapoints related to already-seen proofs which may or may not have been
+/// retained in the outer [`ProofRetainer`].
+///
+/// "target" refers those paths/proofs which are retained by the [`ProofRetainer`], as determined
+/// by its `target` field.
+#[derive(Default, Clone, Debug)]
+struct SeenProofs {
+    last_target_leaf_path: Nibbles,
+    last_nontarget_path: Nibbles,
+    last_nontarget_proof: Vec<u8>,
+}
+
+impl SeenProofs {
+    /// Stores the path of the most recently seen target leaf.
+    fn retain_target_leaf(&mut self, path: &Nibbles) {
+        self.last_target_leaf_path = *path;
+    }
+
+    /// Stores the path and proof of the most recently seen path/proof which were not retained.
+    fn retain_nontarget(&mut self, path: &Nibbles, proof: &[u8]) {
+        self.last_nontarget_path = *path;
+        self.last_nontarget_proof.clear();
+        self.last_nontarget_proof.extend(proof);
+    }
+
+    /// Checks if a branch node is parent to both last target leaf and last non-target node.
+    fn branch_is_parent(&self, path: &Nibbles) -> bool {
+        let l = path.len();
+        self.last_target_leaf_path.len() == l + 1
+            && self.last_nontarget_path.len() == l + 1
+            && self.last_target_leaf_path.starts_with(path)
+            && self.last_nontarget_path.starts_with(path)
```

Comment:
would prefer just using `path.len()` directly
```suggestion
        self.last_target_leaf_path.len() == path.len() + 1
            && self.last_nontarget_path.len() == path.len() + 1
            && self.last_target_leaf_path.starts_with(path)
            && self.last_nontarget_path.starts_with(path)
```